storage/innobase/handler/handler0alter.cc

```diff
@@ -8433,6 +8433,188 @@ innobase_update_foreign_cache(
 	DBUG_RETURN(err);
 }
 
+/** Changes SYS_COLUMNS.PRTYPE for one column.
+@param[in,out]	trx	transaction
+@param[in]	table_name	table name
+@param[in]	tableid	table ID as in SYS_TABLES
+@param[in]	pos	column position
+@param[in]	prtype	new precise type
+@return		boolean flag
+@retval	true	on failure
+@retval false	on success */
+static
+bool
+change_field_versioning_try(
+	trx_t* trx,
+	const char* table_name,
+	const ulint tableid,
+	const ulint pos,
+	const ulint prtype)
+{
+	DBUG_ENTER("change_field_versioning_try");
+
+	pars_info_t* info = pars_info_create();
+
+	pars_info_add_int4_literal(info, "prtype", prtype);
+	pars_info_add_ull_literal(info,"tableid", tableid);
+	pars_info_add_int4_literal(info, "pos", pos);
+
+	dberr_t error = que_eval_sql(info,
+				     "PROCEDURE CHANGE_COLUMN_MTYPE () IS\n"
+				     "BEGIN\n"
+				     "UPDATE SYS_COLUMNS SET PRTYPE=:prtype\n"
+				     "WHERE TABLE_ID=:tableid AND POS=:pos;\n"
+				     "END;\n",
+				     false, trx);
+
+	if (error != DB_SUCCESS) {
+		my_error_innodb(error, table_name, 0);
+		trx->error_state = DB_SUCCESS;
+		trx->op_info = "";
+		DBUG_RETURN(true);
+	}
+
+	DBUG_RETURN(false);
+}
+
+/** Changes fields WITH/WITHOUT SYSTEM VERSIONING property in SYS_COLUMNS.
+@param[in]	ha_alter_info	alter info
+@param[in]	ctx	alter inplace context
+@param[in]	trx	transaction
+@param[in]	table	old table
+@return		boolean flag
+@retval	true	on failure
+@retval false	on success */
+static
+bool
+change_fields_versioning_try(
+	const Alter_inplace_info* ha_alter_info,
+	const ha_innobase_inplace_ctx* ctx,
+	trx_t* trx,
+	const TABLE* table)
+{
+	DBUG_ENTER("change_fields_versioning_try");
+
+	DBUG_ASSERT(ha_alter_info);
+	DBUG_ASSERT(ctx);
+
+	if (!(ha_alter_info->handler_flags & ALTER_COLUMN_UNVERSIONED)){
+		DBUG_RETURN(false);
+	}
+
+	uint virtual_count = 0;
+
+	List_iterator_fast<Create_field> it(
+	    ha_alter_info->alter_info->create_list);
+
+	for (uint i = 0; i < table->s->fields; i++) {
+		const Field* field = table->field[i];
+
+		if (innobase_is_v_fld(field)) {
+			virtual_count++;
+			continue;
+		}
+
+		const Create_field* create_field = NULL;
+		while (const Create_field* cf = it++) {
+			if (cf->field == field) {
+				create_field = cf;
+				break;
+			}
+		}
+		it.rewind();
+		DBUG_ASSERT(create_field);
+
+		if (create_field->versioning
+		    == Column_definition::VERSIONING_NOT_SET) {
+			continue;
+		}
+
+		const dict_table_t* new_table = ctx->new_table;
+		ulint pos = i - virtual_count;
+		const dict_col_t* col = dict_table_get_nth_col(new_table, pos);
+
+		DBUG_ASSERT(!col->vers_sys_start());
+		DBUG_ASSERT(!col->vers_sys_end());
+
+		ulint new_prtype
+		    = create_field->versioning
+			      == Column_definition::WITHOUT_VERSIONING
+			  ? col->prtype & ~DATA_VERSIONED
+			  : col->prtype | DATA_VERSIONED;
+
+		if (change_field_versioning_try(
+			trx, table->s->table_name.str,
+			static_cast<ulint>(new_table->id), pos, new_prtype)) {
```

Comment:
Remove the cast. We must pass all 64 bits of `dict_table_t::id` also on 32-bit systems.

The formula of `pos` looks OK to me. For `SYS_FIELDS.POS` it would encode a prefix length, but `SYS_COLUMNS.POS` is a simple number for non-virtual columns.